lib/syslog/writer.go

```diff
@@ -244,6 +304,51 @@ func (c bufferedConn) Close() error                { return c.conn.Close() }
 func (c bufferedConn) Flush() error                { return c.buf.Flush() }
 func (c bufferedConn) Write(b []byte) (int, error) { return c.buf.Write(b) }
 
+func cachedDialWriter(network string, address string, config *tls.Config, socksProxy string) (w io.Writer, err error) {
+	key := writerCacheKey(network, address, socksProxy)
+	if w, ok := writerCache.get(key); ok {
+		return w, nil
+	}
+	w, err = syncedDialWriter(network, address, config, socksProxy)
+	if err != nil {
+		writerCache.set(key, w)
+	}
+	return w, err
+}
+
+type dialerMutexMap struct {
+	sync.Mutex
+	mu map[string]*sync.Mutex
+}
+
+func (mm *dialerMutexMap) lock(key string) {
+	mm.Lock()
+	defer mm.Unlock()
+	mu, ok := mm.mu[key]
+	if !ok {
+		mu = new(sync.Mutex)
+		mm.mu[key] = mu
+	}
+	mu.Lock()
+}
+
+func (mm *dialerMutexMap) unlock(key string) {
+	if mu, ok := mm.mu[key]; ok {
```

Comment:
Here `mm.mu` is accessed without holding a lock on `mm.Mutex`, this passes the race detector in tests because there are no tests that cause concurrent changes to the map, but in a production setup this will be happening.